devito/passes/iet/mpi.py

```diff
@@ -59,120 +59,116 @@ def _hoist_halospots(iet):
     """
     Hoist HaloSpots from inner to outer Iterations where all data dependencies
     would be honored.
-    """
-
-    # Hoisting rules -- if the retval is True, then it means the input `dep` is not
-    # a stopper to halo hoisting
-
-    def rule0(dep, candidates, loc_dims):
-        # E.g., `dep=W<f,[x]> -> R<f,[x-1]>` and `candidates=({time}, {x})` => False
-        # E.g., `dep=W<f,[t1, x, y]> -> R<f,[t0, x-1, y+1]>`, `dep.cause={t,time}` and
-        #       `candidates=({x},)` => True
-        return (all(i & set(dep.distance_mapper) for i in candidates) and
-                not any(i & dep.cause for i in candidates) and
-                not any(i & loc_dims for i in candidates))
 
-    def rule1(dep, candidates, loc_dims):
-        # A reduction isn't a stopper to hoisting
-        return dep.write is not None and dep.write.is_reduction
+    Example:
+                                   haloupd v[t0]
+    for time                       for time
+      W v[t1]- R v[t0]             W v[t1]- R v[t0]
+      haloupd v[t1]                  haloupd v[t1]
+      R v[t1]                        R v[t1]
+      haloupd v[t0]                  R v[t0]
+      R v[t0]
 
-    rules = [rule0, rule1]
+    """
 
     # Precompute scopes to save time
     scopes = {i: Scope([e.expr for e in v]) for i, v in MapNodes().visit(iet).items()}
 
+    cond_mapper = _create_cond_mapper(iet)
+
     # Analysis
     hsmapper = {}
     imapper = defaultdict(list)
-    for iters, halo_spots in MapNodes(Iteration, HaloSpot, 'groupby').visit(iet).items():
-        for hs in halo_spots:
-            hsmapper[hs] = hs.halo_scheme
 
-            for f, v in hs.fmapper.items():
-                loc_dims = frozenset().union([q for d in v.loc_indices
-                                              for q in d._defines])
+    # Look for parent Iterations of children HaloSpots
+    for iters, halo_spots in MapNodes(Iteration, HaloSpot, 'groupby').visit(iet).items():
+        for i, hs0 in enumerate(halo_spots):
 
-                for n, i in enumerate(iters):
-                    if i not in scopes:
-                        continue
+            # Nothing to do if the HaloSpot is void
+            if hs0.halo_scheme.is_void:
+                continue
 
-                    candidates = [i.dim._defines for i in iters[n:]]
+            for hs1 in halo_spots[i+1:]:
```

Comment:
maybe you could avoid an indentation level but rather considering all pairs

```
from itertools import combinations

...

for hs0, hs1 in combinations(halo_schemes, 2):
    ...
```

Reply:
btw, are we sure that the current code works if there are *2* hoisting opportunities, instead of *1* ?

could we add a test?

Reply:
I am working on this now

Reply:
done